libtenzir/builtins/formats/json.cpp

```diff
@@ -336,695 +175,358 @@ class doc_parser {
+    for (auto doc_it = stream.begin(); doc_it != stream.end();
+         ++doc_it, ++objects_parsed) {
+      if (auto err = doc_it.error()) {
+        diagnostic::warning("{}", error_message(err))
+          .note("skipped invalid JSON at index {}", doc_it.current_index())
+          .emit(*dh);
+        ++diags_emitted;
+        continue;
       }
-    } else {
-      auto [action, slices] = this->handle_selector(doc, json_line, state);
-      switch (action) {
-        case parser_action::parse:
-          break;
-        case parser_action::skip:
-          co_return;
-        case parser_action::yield:
-          TENZIR_ASSERT(slices);
-          for (auto& slice : *slices) {
-            co_yield std::move(slice);
-          }
+      auto doc = *doc_it;
+      if (auto err = doc.error()) {
+        diagnostic::warning("{}", error_message(err))
+          .note("skipped invalid JSON `{}`", truncate(doc_it.source()))
+          .emit(*dh);
+        ++diags_emitted;
+        continue;
       }
-      auto& builder = state.get_active_entry().builder;
-      auto success
-        = doc_parser{json_line, this->ctrl_, lines_processed_, no_infer_, raw_}
-            .parse_object(val.value_unsafe(), builder.record());
-      // After parsing one JSON object it is expected for the result to be at
-      // the end. If it's otherwise then it means that a line contains more than
-      // one object in which case we don't add any data and emit a warning.
-      // It is also possible for a parsing failure to occurr in doc_parser. the
-      // is_alive() call ensures that the first object was parsed without
-      // errors. Calling at_end() when is_alive() returns false is unsafe and
-      // resulted in crashes.
-      if (success and not doc.at_end()) {
-        diagnostic::warning(
-          "encountered more than one JSON object in a single NDJSON line")
-          .note("skips remaining objects in line `{}`", json_line)
-          .emit(this->ctrl_.diagnostics());
-        success = false;
+      auto val = doc.get_value();
+      if (auto err = val.error()) {
+        diagnostic::warning("{}", error_message(err))
+          .note("skipped invalid JSON `{}`", truncate(doc_it.source()))
+          .emit(*dh);
+        ++diags_emitted;
+        continue;
       }
+      auto parser = doc_parser{json_line, *dh, lines_processed_};
+      auto success = parser.parse_object(val.value_unsafe(), builder.record());
       if (not success) {
-        // We already reported the issue.
         builder.remove_last();
+        ++diags_emitted;
+        return;
       }
     }
-    if (auto slices = this->handle_max_rows(state)) {
-      for (auto& slice : *slices) {
-        co_yield std::move(slice);
-      }
+    if (objects_parsed == 0 and diags_emitted == 0) {
+      diagnostic::warning(
+        "NDJSON line did not contain a single valid JSON object")
+        .note("skipped invalid JSON `{}`", truncate(json_line))
+        .emit(*dh);
```

Comment:
```suggestion
      diagnostic::warning(
        "line did not contain a single, valid JSON object")
        .note("skipped invalid JSON `{}`", truncate(json_line))
        .emit(*dh);
```
Not necessarily NDJSON.